### ISSUE TITLE
[RFC]Making sequence-based RL first-class in TorchRL

### ISSUE THREAD

**Issue by theap06** [pytorch/rl#3678]
##Motivation 
TorchRL has all the building blocks for sequence-based RL — SliceSampler, split_trajectories, LSTMModule, TensorDictPrimer, traj_ids — but they are not designed to compose automatically. Users who want to train RNN or sequence-based policies (ACT, Decision Transformer, recurrent PPO) must manually wire them together and get the masking, trajectory key alignment, and hidden state resets right themselves. This RFC proposes making that pipeline explicit, safe, and ergonomic by default.

##Solution

1. Standardize trajectory key defaults
Align SliceSampler's default traj_key to auto-detect ("collector", "traj_ids") when present, falling back to "episode" for backward compatibility. This fixes a silent composability bug with no API break.

2. Add episode_start metadata from collectors
Emit ("collector", "episode_start") — a boolean flag marking the first step of each episode. Downstream code should not have to re-derive episode boundaries from done repeatedly.

3. Sequence sampling with built-in mask and lengths
Extend SliceSampler to return [B, T] TensorDicts with:

("collector", "mask")     # bool [B, T] — False on padded timesteps
("collector", "lengths")  # int  [B]
when strict_length=False. Users should not need to call split_trajectories() manually.

4. Safe hidden state reset for recurrent policies
Recurrent policies currently leak hidden state across episode boundaries unless the user manually configures TensorDictPrimer + is_init. The target behavior: if episode_start or is_init is present, hidden state resets automatically at those positions.

5. Mask-aware losses
Common loss modules should respect ("collector", "mask") if present, so padded timesteps do not contribute to the gradient. Start narrow — behavior cloning / imitation losses first.

**Comment by vmoens** [comment]
Yep keen to integrate all of this.
We have a transform for init tracking called InitTracker. It writes a "is_init" entry. I'd keep that if it's the same as your collector episode start solution.

**Comment by theap06** [comment]
@vmoens Thanks, InitTracker covers the episode boundary tracking I had in mind. I'll use is_init as the key. The remaining scope for this pr would be on aligning SliceSampler's default trajectory key, adding [B, T] sequence sampling with mask/lengths, and making recurrent modules respect is_init for hidden state reset. Does that scope sound right?

**Comment by vmoens** [comment]
Yep! 
Recurrent modules should be ok in the way they deal with is_init, but the auto plumbing needs to be addressed as you mentioned in the issue.

**Comment by elin-bdai** [comment]
Hello! This is great! I’m very interested in this improvement, specifically for recurrent PPO to start. Happy to contribute too.

**Comment by theap06** [comment]
@elin-bda Thanks! I opened a pr regarding the ideas that both myself and vincent thought of. Feel free to leave feedback!

**Comment by theap06** [comment]
@elin-bdai What is your current setup for the recurrent PPO? Is it primarily for robotics control or more general sequence-based RL tasks?

**Comment by vmoens** [comment]
@theap06 
About #3695:
- my thinking is that we don't need masking. Collectors, replay buffers and RNNs can handle sequentially attached trajectories and that should become the norm. Masking / padding is bad. We can store sparse is_init and done states to handle sampling, RNN boundaries and GAE computation.
- the new feature in collectors that makes it possible to yield completed trajectories goes in the same direction, in that it makes the `ndim` arg in replay buffers kind of redundant. Populating 1d buffers with completed trajectories has a slight overhead compared to stacking incomplete multidim trajs like we used to do, but it is way simpler to sample from (and also to write into with many concurrent collectors).
I tried to incorporate this in your PR which kind of contrasts with the statement you made in this issue -- just wanted to make sure we're on the same page regarding unpadded 1d trajectories as first citizens in the lib.

**Comment by theap06** [comment]
@vmoens Makes sense. I agree that we should use a 1D contiguous + sparse is_init/done markers as the primitive and drop pad_output / mask / lengths from the sampler. Before I rework, should I hard-remove or FutureWarning for two releases? And do LSTMModule / GAE already handle 1D sequentially-attached input correctly via is_init / sparse done, or is there a gap I should close in this PR?

**Comment by elin-bdai** [comment]
> [@elin-bdai](https://github.com/elin-bdai) What is your current setup for the recurrent PPO? Is it primarily for robotics control or more general sequence-based RL tasks?

It is for robotics control. We have been using RSL-RL, but we are interested in TorchRL. I will take a look at your PR!

**Comment by vmoens** [comment]
> [@vmoens](https://github.com/vmoens) Makes sense. I agree that we should use a 1D contiguous + sparse is_init/done markers as the primitive and drop pad_output / mask / lengths from the sampler. Before I rework, should I hard-remove or FutureWarning for two releases? And do LSTMModule / GAE already handle 1D sequentially-attached input correctly via is_init / sparse done, or is there a gap I should close in this PR?

I already made these changes. I will give a last look at this before merging, it touches a lot of things
